Fix various issues with the "make buildx" target

thaJeztah opened the pull request:
relates to https://github.com/moby/moby/pull/39340 (and slightly related to https://github.com/moby/moby/issues/40088)

See individual commit messages for a description 😅

thaJeztah committed:
make buildx: fix variable substitution and version

The variables were not substituted, because single-quotes were used.
While at it; change the fixed version/commit to use the actual commit
and version, using git.

before this change:

    make buildx && ./bundles/buildx version
    github.com/docker/buildx ${BUILDX_COMMIT} ${BUILDX_COMMIT}

after this change:

    make buildx && ./bundles/buildx version
    buildx v0.3.0 c967f1d

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
make buildx: fix Makefile version being ignored

The `BUILDX_COMMIT` variable was set as a Make variable,
which isn't exported, and thus not available in scripts,
unless referenced through `$(VAR)` (non-curly-brackets).
As a result `--build-arg BUILDX_COMMIT` did not set the
`BUILDX_COMMIT` build-arg, and the default from the Dockerfile
(`master`) was used instead.

This patch exports the default version that's set in the
Makefile, so that it can be used as a regular environment
variable. The script was also slighly modified to no longer
use the `Make` variable.

In addition, the `buildx` target now calls `buildx version`,
which is useful to confirm if the binary was successfully built
(and with the correct version).

Before:

    rm -f bundles/buildx && make buildx && ./bundles/buildx version
    # => => naming to docker.io/library/moby-buildx:v0.3.0
    github.com/docker/buildx v0.3.1 6db68d0

    # using a make variable:
    rm -f bundles/buildx && make BUILDX_COMMIT=v0.2.1 buildx && ./bundles/buildx version
    # => => naming to docker.io/library/moby-buildx:v0.2.1
    github.com/docker/buildx v0.3.1 6db68d0

    # using an environment variable:
    rm -f bundles/buildx && BUILDX_COMMIT=v0.2.2 make buildx && ./bundles/buildx version
    # => => naming to docker.io/library/moby-buildx:v0.2.2
    github.com/docker/buildx v0.3.1 6db68d0

After:

    # default
    rm -f bundles/buildx && make buildx
    # => => naming to docker.io/library/moby-buildx:v0.3.0
    github.com/docker/buildx v0.3.0 c967f1d

    # using a make variable:
    rm -f bundles/buildx && make BUILDX_COMMIT=v0.2.1 buildx
    # => => naming to docker.io/library/moby-buildx:v0.2.1
    github.com/docker/buildx v0.2.1 0eb2df5

    # using an environment variable:
    rm -f bundles/buildx && BUILDX_COMMIT=v0.2.2 make buildx
    # => => naming to docker.io/library/moby-buildx:v0.2.2
    github.com/docker/buildx v0.2.2 ab5fe3d

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
make buildx: make version in Dockerfile leading

This patch removes the `BUILDX_COMMIT` make variable. With the
make variable removed, it no longer "masks" environment variables,
and there is no longer a need to export the variable.

A side effect of this change, is that (by default), the buildx
image is tagged as `moby-buildx:latest`. This likely isn't a
problem, because the build-cache would still be preserved in
intermediate images. Having the image tagged as `:latest` also
makes cleaning up easier (without having to remove the image
for each version tagged.

Otherwise, the behavior remains the same as before:

    # default
    rm -f bundles/buildx && make buildx
    # => => naming to docker.io/library/moby-buildx:latest
    github.com/docker/buildx v0.3.0 c967f1d

    # using a make variable:
    rm -f bundles/buildx && make BUILDX_COMMIT=v0.2.1 buildx
    # => => naming to docker.io/library/moby-buildx:v0.2.1
    github.com/docker/buildx v0.2.1 0eb2df5

    # using an environment variable:
    rm -f bundles/buildx && BUILDX_COMMIT=v0.2.2 make buildx
    # => => naming to docker.io/library/moby-buildx:v0.2.2
    github.com/docker/buildx v0.2.2 ab5fe3d

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
make buildx: remove BUILDX_DOCKERFILE make var

This simplifies the makefile a bit, while preserving
the functionality. Using a non-existing Dockerfile
to demonstrate:

    make buildx
    Successfully tagged moby-buildx:latest
    92059305df7371f8b5b3638d4d405d49ff909031a7bc6d2f515cb0a0df03c2f4
    github.com/docker/buildx v0.3.0 c967f1d

    make BUILDX_DOCKERFILE=foo buildx
    BUILDX_DOCKERFILE=foo buildx
    unable to prepare context: unable to evaluate symlinks in Dockerfile path: lstat /Users/sebastiaan/go/src/github.com/docker/docker/foo: no such file or directory

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
make buildx: clean-up comment and fix help-text

Un-indent the comment, so that it doesn't get printed by
the shell script (moved it above the target, as it looked
slightly less cluttered)

Also fixed the "help" comment, so that it shows up in
`make help`, and removed the un-needed dummy `buildx:` target.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
make buildx: use multi-stage to make tagged image smaller

The build-stage would still be in the local cache (and can
be cleaned up with `docker system prune`) but the tagged
image (which usually wouldn't be removed) will take up
less space now.

Note that

 - the binary is not statically linked, so we cannot use
   a "from scratch" image
 - in cases where the binary is cross-compiled (e.g.
   on a non-linux machine), the image itself is not
   really useful (we may want to consider not tagging
   the image in that situation)

Before:

    REPOSITORY      TAG     IMAGE ID      CREATED         SIZE
    moby-buildx     latest  c9b2af465baf  7 minutes ago   1.71GB

After:

    REPOSITORY      TAG     IMAGE ID      CREATED         SIZE
    moby-buildx     latest  345501e2df0a  2 minutes ago   820MB

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
ping @cpuguy83 ptal

thaJeztah commented:
Hm... interesting; am I dropping an env-var here, causing it to fall back to `go env` ?

```
 + make bundles/buildx
 mkdir bundles
 docker build -f ${BUILDX_DOCKERFILE:-Dockerfile.buildx} -t "moby-buildx:${BUILDX_COMMIT:-latest}" \
 	--build-arg BUILDX_COMMIT \
 	--build-arg BUILDX_REPO \
 	--build-arg GOOS=$(if [ -n "" ]; then echo ; else go env GOHOSTOS || uname | awk '{print tolower($0)}' || true; fi) \
 	--build-arg GOARCH=$(if [ -n "" ]; then echo ; else go env GOHOSTARCH || true; fi) \
 	. && \
 	id=$(docker create moby-buildx:${BUILDX_COMMIT:-latest}); \
 	if [ -n "${id}" ]; then docker cp ${id}:/usr/bin/buildx bundles/buildx && touch bundles/buildx; docker rm -f ${id}; fi \
 && bundles/buildx version
 /bin/sh: 4: go: not found
 /bin/sh: 5: go: not found
```

thaJeztah commented on Makefile:
```diff
@@ -247,21 +247,19 @@ swagger-docs: ## preview the API documentation
 		--build-arg BUILDX_COMMIT \
 		--build-arg BUILDX_REPO \
 		--build-arg GOOS=$$(if [ -n "$(GOOS)" ]; then echo $(GOOS); else go env GOHOSTOS || uname | awk '{print tolower($$0)}' || true; fi) \
 		--build-arg GOARCH=$$(if [ -n "$(GOARCH)" ]; then echo $(GOARCH); else go env GOHOSTARCH || true; fi) \
```
oh! `$(GOOS)` and `$(GOARCH)` are Make variables as well 🤔. I wonder why it worked before though

thaJeztah replied:
Perhaps I need to borrow this snippet from containerd; https://github.com/containerd/containerd/blob/036db34f37617d6a02f07b04fc2d35a91732c6e3/Makefile#L27-L52

Or set it in the Jenkinsfile

thaJeztah replied:
Ah! LOL.. then it worked, but didn't work because now we _actually_ used `v0.3.0`, which doesn't support s390x/ppc64le; The `copy` image use in v0.3.0 does not have s390x/ppc64le;

```
#3 docker-image://docker.io/tonistiigi/copy:v0.1.3@sha256:e57a3b4d6240f55ba...
#3 digest: sha256:150c626a5355601b2c71794d6e4a9af463e674e78c8f6cd2b0839485d62d8c5f
#3 name: "docker-image://docker.io/tonistiigi/copy:v0.1.3@sha256:e57a3b4d6240f55bac26b655d2cfb751f8b9412d6f7bb1f787e946391fb4b21b"
#3 started: 2019-10-14 16:57:58.920184883 +0000 UTC
#3 completed: 2019-10-14 16:57:58.920265067 +0000 UTC
#3 duration: 80.184
#3 started: 2019-10-14 16:57:58.920324378 +0000 UTC
#3 completed: 2019-10-14 16:57:58.922283274 +0000 UTC
#3 duration: 1.958896ms
#3 error: "no match for current platform linux/s390x in manifest sha256:e57a3b4d6240f55bac26b655d2cfb751f8b9412d6f7bb1f787e946391fb4b21b: not found"
#3 resolve docker.io/tonistiigi/copy:v0.1.3@sha256:e57a3b4d6240f55bac26b655d2cfb751f8b9412d6f7bb1f787e946391fb4b21b done
```

thaJeztah replied:
See my other comment; it's because this PR adds a `COPY`, and therefore `docker.io/tonistiigi/copy:v0.1.3` has to be pulled, which is only available for `amd64` and `armv7`

thaJeztah committed:
make buildx: update to buildx v0.3.1

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

cpuguy83 commented on Makefile:
```diff
@@ -35,6 +35,49 @@ export VALIDATE_REPO
 export VALIDATE_BRANCH
 export VALIDATE_ORIGIN_BRANCH
 
+ifneq "$(strip $(shell command -v go 2>/dev/null))" ""
```
This all seems a bit overkill.
What are we trying to solve here?

thaJeztah replied:
@cpuguy83 I ran into this problem;

```
 /bin/sh: 4: go: not found
 /bin/sh: 5: go: not found
```

thaJeztah replied:
Likely because the `GOOS` and `GOARCH` make variables were not set

cpuguy83 replied:
Yeah but doesn't what we already have solve that? And in the case no `go` the user just needs to set `GOOS` and `GOARCH` themselves.

thaJeztah replied:
But I just discovered what the actual problem was; we were now actually using buildx v0.3.0, so it tried to pull an image that doesn't exist for s390x/power. The `go not found` was (I think) just the output leaking from the `go env GOHOSTOS || uname` line; I'll remove that one commit

thaJeztah replied:
Oh, actually.. (sleepy head); it's because s390x/power are now using buildkit to build the image 🤔

thaJeztah commented:
Yo, Jenkins, where did you go?

thaJeztah committed:
Jenkinsfile: use buildx on s390x and ppc64le

Commit 1be2cc2568ddd52a3b27d607a9c9f7c9347d3c50 updated the
Makefile to force the use of BuildKit, if `USE_BUILDX` was
not set.

As a side-effect, Jenkins now started using BuildKit on
s390x and ppc64le as well, because it overwrote the
`DOCKER_BUILDKIT=0` that was set.

This commit forces the use of buildx on s390x and ppc64le.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on Dockerfile.buildx:
```diff
@@ -1,14 +1,27 @@
+ && go build -ldflags "${ldflags}" -o /usr/bin/buildx ./cmd/buildx
+
+FROM golang:${GO_VERSION}-stretch
+COPY --from=build /usr/bin/buildx /usr/bin/buildx
```
Ah! 💡 I know why it failed in this branch; before this, there was no `COPY` in the Dockerfile, and because the `COPY` was added, it started downloading the `docker.io/tonistiigi/copy:v0.1.3` image. Apparently, that one is only available for amd64 ?

thaJeztah replied:
Yup; that one is only amd64, and armv7

cpuguy83 replied:
What's the reasoning for doing this second image?
This only produces a runnable image if the specified GOOS/GOARCH == the daemon's GOOS/GOARCH

In effect this is creating a 2nd image and taking even more space since buildx is duplicated.

thaJeztah replied:
The first stage won't be tagged, so can be garbage collected, and only the second one is preserved; the final image is roughly 870MB whereas without that stage, we're tagging an image that's 1.71GB; see the commit message https://github.com/moby/moby/commit/39f1a4ab44775428e95508ef8f186a81d60dc907

thaJeztah commented:
OK, this builds now, but I'll amend one commit message, and re-order commits (enabling BUILDX _before_ the commit that adds the `COPY`)

kolyshkin commented on Dockerfile.buildx:
```diff
@@ -10,5 +10,12 @@ ARG GOOS
-ENTRYPOINT ["/usr/bin/buildx"]
\ No newline at end of file
+RUN GOOS="${GOOS}" GOARCH="${GOARCH}" BUILDX_COMMIT="${BUILDX_COMMIT}" \
+ && export version="$(git describe --tags)" \
```
there should not be `&&` in here as what is before (var assignments) is not a command.

kolyshkin commented on Dockerfile.buildx:
```diff
@@ -10,5 +10,12 @@ ARG GOOS
\ No newline at end of file
+RUN GOOS="${GOOS}" GOARCH="${GOARCH}" BUILDX_COMMIT="${BUILDX_COMMIT}" \
+ && export version="$(git describe --tags)" \
+ && export commit="$(git rev-parse --short HEAD)" \
```
1 `export` allows for multiple var assignments
2 you don't need to export any variables as you're using them only here ("export" means make them available for child processes, and I don't see how it's necessary here)

kolyshkin replied:
Oh I see why you're using export... so you would use the vars later. Well, I still don't like the way it looks :(

thaJeztah replied:
Let me try again; I think I tried without export, and for some reason, things failed.

kolyshkin commented on Makefile:
```diff
@@ -252,16 +252,19 @@ else
 	# The idea here being that if buildx is being used, it's because buildkit is not supported natively
-	docker build -f $(BUILDX_DOCKERFILE) -t "moby-buildx:$(BUILDX_COMMIT)" \
+	docker build -f $(BUILDX_DOCKERFILE) -t "moby-buildx:$${BUILDX_COMMIT:-latest}" \
 		--build-arg BUILDX_COMMIT \
```
So, why not just use `--build-arg BUILDX_COMMIT=$(BUILDX_COMMIT)` instead?

thaJeztah replied:
If possible, I like to have the build-args take their value from the environment. That way if no env-var is set, the value from the Dockerfile is leading. Putting the value _after_ means if it's not set, it will be `--build-arg=`, which is equivalent to `--build-arg=""`, which will override the value in the Dockerfile

kolyshkin replied:
ok, makes sense

kolyshkin commented on Makefile:
```diff
@@ -252,16 +252,19 @@ else
 		. && \
-		id=$$(docker create moby-buildx:$(BUILDX_COMMIT)); \
-		if [ -n "$${id}" ]; then docker cp $${id}:/usr/bin/buildx $@ && touch $@; docker rm -f $${id}; fi
+		id=$$(docker create moby-buildx:$${BUILDX_COMMIT:-latest}); \
```
OK, so using `&&` from makefile is redundant. If any command fails, make stops right there. I guess this practice comes from Dockerfile `RUN` use, but makefile is different and if we can make things simpler, let's do it.

In other words,
```
	docker build ....
	id=$$(docker ...); if [ -n "$${id}" ]; ....
	$@ version
```

thaJeztah replied:
Ok I rewrote it a bit; also put the `if` over a couple of lines, but let me know if that's too verbose.

cpuguy83 commented on Makefile:
```diff
@@ -247,21 +247,25 @@ swagger-docs: ## preview the API documentation
-buildx: bundles/buildx # build buildx cli tool
-else
-buildx:
+buildx: bundles/buildx ## build buildx cli tool
```
In https://github.com/moby/moby/pull/40091 buildx is a required pre-req for all builds.
What happens if we don't define `buildx:`?

thaJeztah replied:
It should be covered by the `.PHONY buildx`

```bash
mkdir foo && cd foo

echo ".PHONY: buildx" > Makefile

make buildx
# make: Nothing to be done for `buildx'.

echo $?
# 0

make nonexistingtarget
# make: *** No rule to make target `nonexistingtarget'.  Stop.

echo $?
2
```

kolyshkin commented on Dockerfile.buildx:
```diff
@@ -10,5 +10,10 @@ ARG GOOS
+    -X \"github.com/docker/buildx/version.Version=$(git describe --tags)\" \
+    -X \"github.com/docker/buildx/version.Revision=$(git rev-parse --short HEAD)\" \
+    -X \"github.com/docker/buildx/version.Package=github.com/docker/buildx\"" \
+ && go build -ldflags "${ldflags}" -o /usr/bin/buildx ./cmd/buildx
```
`&&` looks artificial here as we don't have any command to succeed before it.

I also see a repetition of `github.com/docker/buildx` string; why don't we do this

```shell
  pkg="github.com/docker/buildx"; \
  ldflags="\
    -X \"${pkg}/version.Version=$(git describe --tags)\" \
    -X \"${pkg}/version.Revision=$(git rev-parse --short HEAD)\" \
    -X \"${pkg}/version.Package=${pkg}\""; \
  echo go build -ldflags "${ldflags}" -o /usr/bin/buildx ./cmd/buildx
```

I checked it works in both `bash` and `dash`, and even passes shellcheck

thaJeztah replied:
> && looks artificial here as we don't have any command to succeed before it.

arf. that's me messing up rebasing

> I also see a repetition of github.com/docker/buildx string; why don't we do this

👍 yup, I like that

kolyshkin commented on Dockerfile.buildx:
```diff
@@ -1,9 +1,12 @@
 RUN git clone "${BUILDX_REPO}" /buildx
 WORKDIR /buildx
-ARG BUILDX_COMMIT=master
+ARG BUILDX_COMMIT
```
Side note: docs at https://github.com/moby/moby/issues/37345#issuecomment-400245466 helped me to review this :)))

thaJeztah replied:
Funny you reference it as "docs". I should try to find some time to incorporate it in the docs though

thaJeztah commented:
Updated with the suggested changes

cpuguy83 reviewed (approved):
LGTM

kolyshkin commented on Makefile:
```diff
@@ -252,16 +252,25 @@ else
+	id=$$(docker create moby-buildx:$${BUILDX_COMMIT:-latest}); \
+	if [ -n "$${id}" ]; then \
+		docker cp $${id}:/usr/bin/buildx $@; \
+		touch $@; \
```
OK, `&&` is still needed before `touch`, since otherwise we'll end up having an empty file in case `docker cp` fails, and this is not what we want here.

kolyshkin replied:
IOW,

`docker cp $${id}:/usr/bin/buildx $@ && touch $@; \`

otherwise all the same

kolyshkin replied:
in fact it's OK as it is since with empty file, `$@ version` below will fail anyway so up to you

thaJeztah replied:
Ah, hm, yeah, overlooked that; I'll fix it to be correct

thaJeztah commented on Dockerfile.buildx:
```diff
@@ -10,5 +10,12 @@ ARG GOOS
+    ldflags="\
+      -X \"${pkg}/version.Version=$(git describe --tags)\" \
+      -X \"${pkg}/version.Revision=$(git rev-parse --short HEAD)\" \
+      -X \"${pkg}/version.Package=github.com/docker/buildx\" \
```
Ah, perhaps we should change to just `buildx`, so that it shows as;

```
buildx version
buildx v0.3.0 c967f1d
```

instead of

```
buildx version
github.com/docker/buildx v0.3.0 c967f1d
```

thaJeztah commented:
@kolyshkin updated 👍

kolyshkin commented on Dockerfile.buildx:
```diff
@@ -10,5 +10,12 @@ ARG GOOS
+    ldflags="\
+      -X \"${pkg}/version.Version=$(git describe --tags)\" \
+      -X \"${pkg}/version.Revision=$(git rev-parse --short HEAD)\" \
+      -X \"${pkg}/version.Package=buildx\" \
```
hmm. shouldn't it be `-X \"${pkg}/version.Package=${pkg}\" \`?

thaJeztah replied:
See https://github.com/moby/moby/pull/40089#discussion_r337160075

This way, it looks like

```
buildx version
buildx v0.3.0 c967f1d
```

instead of

```
buildx version
github.com/docker/buildx v0.3.0 c967f1d
```

thaJeztah commented:
rebased to kick CI (master was broken because of a linting issue)

kolyshkin reviewed (approved):
LGTM

thaJeztah commented:
argh; failing on a flaky test;

```
 === Failed
[2019-10-21T23:02:53.486Z] === FAIL: amd64.integration-cli TestDockerSwarmSuite/TestAPISwarmNodeDrainPause (72.72s)
[2019-10-21T23:02:53.486Z]     --- FAIL: TestDockerSwarmSuite/TestAPISwarmNodeDrainPause (72.72s)
[2019-10-21T23:02:53.486Z]         daemon.go:26: Creating a new daemon at: "/go/src/github.com/docker/docker/bundles/test-integration/TestDockerSwarmSuite/TestAPISwarmNodeDrainPause"
[2019-10-21T23:02:53.486Z]         daemon.go:26: Creating a new daemon at: "/go/src/github.com/docker/docker/bundles/test-integration/TestDockerSwarmSuite/TestAPISwarmNodeDrainPause"
[2019-10-21T23:02:53.486Z]         check_test.go:358: [d48494135b48e] joining swarm manager [d1822cb98a9bc]@0.0.0.0:2477, swarm listen addr 0.0.0.0:2478
[2019-10-21T23:02:53.486Z]         docker_api_swarm_node_test.go:106: timeout hit after 1m0s: output: "b99585122cdb\ncdbb41970aad\n"
[2019-10-21T23:02:53.486Z]         daemon.go:83: assertion failed: error is not nil: exit status 1
```
